### DISCUSSION TITLE
Single Precision Output of 3D Rectilinear Data

### DISCUSSION THREAD

**Discussion by wilfonba** [LLNL/Silo#310] [Q&A] [answered]
I am trying to add single-precision output of Silo HDF5 files to an existing Fortran HPC codebase, and I've been unable to get it working with 3D rectilinear data. The 2D case works as expected, but when I try a 3D case I get an empty outline when trying to visualize the mesh in VisIt and "The Contour plot of variable "alpha1" yielded no data." when trying to display a variable in the silo file. The relevant code is available at https://github.com/belericant/MFC/blob/modular-fp/src/post_process/m_data_output.fpp, and I've attached a root and p0 silo file for reference as well. I'm posting to see if anyone else has encountered a similar issue when using single-precision output with Fortran, or if this is a possible bug since the 2D and 3D code is so similar. Any input would be greatly appreciated. Thanks!
[collection_0.txt](https://github.com/LLNL/Silo/files/11437483/collection_0.txt)
[0.txt](https://github.com/LLNL/Silo/files/11437485/0.txt)

**Comment by markcmiller86** [comment]
Quick couple of thoughts...

1. For single precision output, I hope you didn't wind up writing any special code for that and just call [`DBForceSingle()`](https://wci.llnl.gov/sites/wci/files/2020-08/LLNL-SM-654357.pdf?#page=42) prior to executing your Silo output code blocks.
2. For 3D case, I'd bet you have a situation where your entire dataset is *embedded* in *ghost* zones. You can maybe confirm by using VisIt's **Inverse Ghost Zone** operator...which makes non-ghost zones ghost and ghost-zones non-ghost. If you wind up with an image, that is the problem. For domains on exterior boundaries, you need to take care that you are ensuring only the portion of those domains that have ghost zones are actually ghosted in the calls to `DBPutQuadmesh()` (controled by `DB_HI_OFFSET` and `DB_LO_OFFSET` options in the associated optlist).

**Reply by wilfonba** [reply]
Thanks for the quick response. Since the code is in Fortran I wasn't able to use `DBForceSingle()`. The code I added serves to cast arrays of doubles to floats, and then make the appropriate calls to `DBPUTQM` and `DBPUTQV1`.

I tried the Visit trick, and after applying the inverse ghost zone operator to the mesh 'rectilinear_grid' I get "The Mesh plot of variable "rectilinear_grid" yielded no data.". One additional note is that when visualizing in Paraview, only a line is shown in the rendering window.

**Reply by markcmiller86** [reply]
Alright...I'll take a closer look at your files tomorrow.

**Comment by markcmiller86** [comment]
It looks like you attached output from `h5dump` here. Are there tools that produce an hdf5 file from output from `h5dump`? I am not aware of any. So, can you instead attach the actual binary silo files here? You will need to gzip compress them to attach them.

**Reply by wilfonba** [reply]
This should contain a 5x5x5 mesh and 8 scalar cell datasets.

[silo.tar.gz](https://github.com/LLNL/Silo/files/11443925/silo.tar.gz)

**Reply by markcmiller86** [reply]
So, this is the coordinate data I see in the file,

```
coords[0]                      =    {{0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0}, {0, 0.8, 1.6, 2.4, 3.2, 4}}
```

This defines a degenerate 3D mesh. I think VisIt has having trouble setting view parameters due to this and so you get kind of an odd view.

Are you trying to define a 1D path in 3D space or a bonified 3D mesh of hexahedra? If the latter, then you cannot have *colinear* coordinates and then have any two successive values in a coordinate array be numerically identical.

**Reply by markcmiller86** [reply]
Also, my apologies about possibly using `DBForceSingle()`. I had forgotten...thats really only for consumers of Silo files. They can set that and then read into memory only 32 bit data even in 64 bit floats are in the file. I've long had the intention of extending that to the write interface so that callers can pass 64 bit float data and wind up with only 32 bit float data in the file. But, with the advent of lossy compressors like ZFP, the value in providing that feature is diminishing.

**Reply by wilfonba** [reply]
Ah, the mesh issue is a result of me trying to debug by setting the z coordinates equal to the x coordinates. These files should be corrected. Apologies for the confusion.

[silo.tar.gz](https://github.com/LLNL/Silo/files/11446221/silo.tar.gz)

**Reply by markcmiller86** [reply, marked as answer]
I opened `p0/0.silo` in most recent attachment and am still seeing the same invalid coordinate arrays

**Reply by wilfonba** [reply]
Ah, I just checked what was being passed to the silo routines and I had a bug in my code in which the x and y values were not being cast to single precision when there were cells in the z direction due to a silly logic error. Thank you very much for your help in tracking this down! It is greatly appreciated.

**Reply by markcmiller86** [reply]
Ok, I opened with `browser` in *writable* mode (e.g. `-W` command line option) and adjust coordinate array to look like...

```
coords[0]                      =    {{0, 0.8, 1.6, 2.4, 3.2, 4}, {0, 0.8, 1.6, 2.4, 3.2, 4}, {0, 0.8, 1.6, 2.4, 3.2, 4}}

```

[0.silo.gz](https://github.com/LLNL/Silo/files/11446270/0.silo.gz)

And, when I do, I get something VisIt plots.

**Reply by wilfonba** [reply]
Thanks again!
